plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUClass.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiIdentifier
+import org.jetbrains.kotlin.asJava.classes.KtLightClass
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.uast.*
+import org.jetbrains.uast.internal.acceptList
+import org.jetbrains.uast.visitor.UastVisitor
+
+sealed class AbstractFirKotlinUClass(
+    givenParent: UElement?
+) : FirKotlinAbstractUElement(givenParent), UClass, UAnchorOwner {
+    override val uAnnotations: List<UAnnotation>
+        get() {
+            // TODO: Not yet implemented
+            return emptyList()
+        }
+
+    override val uastDeclarations: List<UDeclaration> by lz {
+        mutableListOf<UDeclaration>().apply {
+            addAll(fields)
+            addAll(initializers)
+            addAll(methods)
+            addAll(innerClasses)
+        }
+    }
+
+    open val ktClass: KtClassOrObject?
+        get() = (psi as? KtLightClass)?.kotlinOrigin
+
+    override val uastSuperTypes: List<UTypeReferenceExpression>
+        get() = ktClass?.superTypeListEntries.orEmpty().mapNotNull { it.typeReference }.map {
+            FirKotlinUTypeReferenceExpression(it, this)
+        }
+
+    // TODO: delegateExpressions
+
+    override fun accept(visitor: UastVisitor) {
+        if (visitor.visitClass(this)) return
+        // TODO: delegate expressions
+        uAnnotations.acceptList(visitor)
+        uastDeclarations.acceptList(visitor)
+        visitor.afterVisitClass(this)
+    }
+
+    companion object {
+        fun create(psi: KtLightClass, givenParent: UElement?): UClass = when (psi) {
+            // TODO: PsiAnonymousClass
+            // TODO: KtLightClassForScript
+            else -> FirKotlinUClass(psi, givenParent)
+        }
+    }
+}
+
+class FirKotlinUClass(
```

Comment:
Should we replace `Fir` in these names `Kt` or `Hl`?

Reply:
On one hand, I think it's a fair point. On the other hand, FIR stuff are still used under the hood, e.g., for `javaPsi`, any generated light-class symbols are `FirLight*`.